test/test.c

```diff
@@ -32058,6 +32179,882 @@ test_KeyValueMirrorCrossDomains(void)
+
+    // start 5 instances of the basic service.
+    for (i = 0; i < NUM_BASIC_MICRO_SERVICES; i++)
+    {
+        snprintf(buf, sizeof(buf), "Start microservice #%d: ", i);
+        test(buf);
+        testCond(NULL == micro_AddService(&svcs[i], nc, &cfg));
+    }
+
+    // Now send 50 requests.
+    test("Send 50 requests: ");
+    for (i = 0; i < 50; i++)
+    {
+        s = natsConnection_Request(&reply, nc, "svc.do", NULL, 0, 1000);
+        if (NATS_OK != s)
+            FAIL("Unable to send request");
+        natsMsg_Destroy(reply);
+    }
+    testCond(NATS_OK == s);
+   
+    // Destroy the services in the reverse order to properly unwind the
+    // callbacks. At some point this needs to be fixed so that the services
+    // exclude themselves from a chain, rather than setting previous, etc.
+    for (i = NUM_BASIC_MICRO_SERVICES - 1; i >= 0; i--)
+    {
+
+        microService_Destroy(svcs[i]);
+    }
+    NATS_FREE(svcs);
+
+    natsConnection_Destroy(nc);
+    natsOptions_Destroy(opts);
+    _destroyDefaultThreadArgs(&arg);
+    _stopServer(serverPid);
+}
+
+static void
+test_MicroServiceStopsOnClosedConn(void)
+{
+    natsStatus s;
+    natsConnection *nc = NULL;
+    natsOptions *opts = NULL;
+    natsPid serverPid = NATS_INVALID_PID;
+    struct threadArg arg;
+    microService *m = NULL;
+    microServiceConfig cfg = {
+        .Name = "test",
+        .Version = "1.0.0",
+    };
+    natsMsg *reply = NULL;
+
+    s = _createDefaultThreadArgsForCbTests(&arg);
+    if (s == NATS_OK)
+        opts = _createReconnectOptions();
+
+    if ((opts == NULL) || (natsOptions_SetURL(opts, NATS_DEFAULT_URL) != NATS_OK))
+    {
+        FAIL("Unable to setup test for MicroConnectionEvents!");
+    }
+
+    serverPid = _startServer("nats://127.0.0.1:4222", NULL, true);
+    CHECK_SERVER_STARTED(serverPid);
+
+    test("Connect for microservice: ");
+    testCond(NATS_OK == natsConnection_Connect(&nc, opts));
+
+    test("Start microservice: ");
+    testCond(NULL == micro_AddService(&m, nc, &cfg));
+
+    test("Test microservice is running: ");
+    testCond(!microService_IsStopped(m))
+
+    test("Test microservice is responding to PING: ");
+    testCond(NATS_OK == natsConnection_RequestString(&reply, nc, "$SRV.PING.test", "", 500));
+    natsMsg_Destroy(reply);
+    reply = NULL;
+
+    test("Stop microservice: ");
+    testCond(NULL == microService_Stop(m))
+
+    test("Test microservice is not running: ");
+    testCond(microService_IsStopped(m))
+
+    test("Test microservice is not responding to PING: ");
+    testCond(NATS_OK != natsConnection_RequestString(&reply, nc, "$SRV.PING.test", "", 500));
+
+    test("Destroy microservice (cleanup): ");
+    testCond(NULL == microService_Destroy(m))
+
+    test("Start microservice again: ");
+    testCond(NULL == micro_AddService(&m, nc, &cfg));
+
+    test("Close the connection: ");
+    testCond(NATS_OK == natsConnection_Drain(nc));
```

Comment:
Looks like there is an issue here this blocks for 60 seconds or more and the test times out, causing other issues later.